MekHQ/src/mekhq/campaign/mission/atb/AtBScenarioFactory.java

```diff
@@ -42,228 +55,274 @@
-    						lList.add(l);
-    					}
-    				}
-    				if (lList.size() == 0) {
-    					for (Lance l : lances.values()) {
-    						if (l.getMissionId() == m.getId() && l.isEligible(c)) {
-    							lList.add(l);
-    						}
-    					}
-    				}
-    				if (lList.size() == 0) {
-    					for (Lance l : lances.values()) {
-    						if (l.isEligible(c)) {
-    							lList.add(l);
-    						}
-    					}
-    				}
-    				if (lList.size() > 0) {
-    					Lance lance = Utilities.getRandomItem(lList);
-    					AtBScenario scenario = AtBScenarioFactory.createScenario(c, lance, AtBScenario.BASEATTACK, false,
-    							Lance.getBattleDate(c.getCalendar()));
-    					for (int i = 0; i < sList.size(); i++) {
-    						if (sList.get(i).getLanceForceId() == lance.getForceId()) {
-    							sList.set(i, scenario);
-    							break;
-    						}
-    					}
-    					if (!sList.contains(scenario)) {
-    						sList.add(scenario);
-    					}
-    				} else {
-    					//TODO: What to do if there are no lances assigned to this contract?
-    				}
-				}
-			}
-		}
-
-		/* Sort by date and add to the campaign */
-		sList.sort(Comparator.comparing(Scenario::getDate));
-		for (AtBScenario s : sList) {
-			c.addScenario(s, c.getMission(s.getMissionId()));
-			s.setForces(c);
-		}
-	}
+        }
+    }
+
+    /**
+     * Iterate through the list of lances and make a battle roll for each,
+     * then sort them by date before adding them to the campaign.
+     * Contracts with enemy morale level of invincible have a base attack
+     * (defender) battle each week. If there is a base attack (attacker)
+     * battle, that is the only one for the week on that mission.
+     *
+     * Note that this handles having multiple missions at the same time
+     * @param c the campaign for which to generate scenarios
+     */
+    public static void createScenariosForNewWeek(Campaign c) {
+        Hashtable<Integer, Lance> lances = c.getLances();
+
+        AtBContract atbContract;
+        List<AtBScenario> sList;
+        List<Integer> assignedLances = new ArrayList<>();
+        List<Integer> dontGenerateForces;
+        boolean hasBaseAttack;
+        boolean hasBaseAttackAttacker;
+
+        // Determine active missions
+        for (Mission mission : c.getMissions()) {
+            if (!mission.isActive() || !(mission instanceof AtBContract) ) {
+                continue; //if not active or an AtBContract, we don't care about the mission
+            }
+
+            //region Value Initialization
+            atbContract = (AtBContract) mission;
+            sList = new ArrayList<>();
+            dontGenerateForces = new ArrayList<>();
+            hasBaseAttack = false;
+            hasBaseAttackAttacker = false;
+            //endregion Value Initialization
+
+            //region Current Scenarios
+            // Determine active scenarios
```

Comment:
Let's put in a little more detailed commentary about why we're looping through all the scenarios in the current contract. I was eventually able to figure it out but it took me a couple of minutes.

Reply:
Will do, forgot that I've been thinking and looking at this for a few weeks now and didn't comment it well.